File: syncloud/src/main/java/org/syncloud/android/core/redirect/RedirectService.java
package org.syncloud.android.core.redirect;

import com.fasterxml.jackson.databind.ObjectMapper;

import org.apache.http.NameValuePair;
import org.apache.http.message.BasicNameValuePair;
import org.apache.log4j.Logger;
import org.syncloud.android.core.common.SyncloudException;
import org.syncloud.android.core.common.WebService;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static org.syncloud.android.core.common.jackson.Jackson.createObjectMapper;

public class RedirectService implements IUserService {

    public static String getApiUrl(String mainDomain) {
        return "https://api."+mainDomain;
    }

    private static Logger logger = Logger.getLogger(RedirectService.class);

    private WebService webService;

    public RedirectService(String apiUrl) {
        this.webService = new WebService(apiUrl);
    }

    private static ObjectMapper mapper = createObjectMapper();

    public org.syncloud.android.core.redirect.model.User getUser(String email, String password) {
        List<NameValuePair> parameters = new ArrayList<NameValuePair>();
        parameters.add(new BasicNameValuePair("email", email));
        parameters.add(new BasicNameValuePair("password", password));

        String json = webService.execute("GET", "/user/get", parameters);

        try {
            org.syncloud.android.core.redirect.model.UserResult restUser = mapper.readValue(json, org.syncloud.android.core.redirect.model.UserResult.class);
            return restUser.data;
        } catch (IOException e) {
            String message = "Failed to deserialize json";
            logger.error(message+" "+json, e);
            throw new SyncloudException(message);
        }
    }

    public org.syncloud.android.core.redirect.model.User createUser(String email, String password) {
        List<NameValuePair> parameters = new ArrayList<NameValuePair>();
        parameters.add(new BasicNameValuePair("email", email));
        parameters.add(new BasicNameValuePair("password", password));

        String json = webService.execute("POST", "/user/create", parameters);

        try {
            org.syncloud.android.core.redirect.model.UserResult restUser = mapper.readValue(json, org.syncloud.android.core.redirect.model.UserResult.class);
            return restUser.data;
        } catch (IOException e) {
            String message = "Failed to deserialize json";
            logger.error(message+" "+json, e);
            throw new SyncloudException(message);
        }
    }

}

File: syncloud/src/main/java/org/syncloud/android/core/redirect/RedirectService.kt
package org.syncloud.android.core.redirect

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.apache.log4j.Logger
import org.syncloud.android.core.common.SyncloudException
import org.syncloud.android.core.common.WebService
import org.syncloud.android.core.redirect.model.User
import org.syncloud.android.core.redirect.model.UserCredentials
import org.syncloud.android.core.redirect.model.UserResult
import java.io.IOException

class RedirectService(private val mainDomain: String, private val webService: WebService) : IUserService {

    private val logger = Logger.getLogger(RedirectService::class.java)
    private val mapper = jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    override fun getUser(email: String, password: String): User? {
        val url = "https://api.$mainDomain/user"
        val requestJson = mapper.writeValueAsString(UserCredentials(email, password))
        val json = webService.post(requestJson, url)
        return try {
            val restUser = mapper.readValue<UserResult>(json)
            restUser.data
        } catch (e: IOException) {
            val message = "Failed to deserialize json"
            logger.error("$message $json", e)
            throw SyncloudException(message)
        }
    }
}